[BUG]: `:v0.7.x` ARM images all fail `python/unit/test/mesh/test_higher_order_mesh.py`

### Summarize the issue

In our ARM images some gmsh-based tests create meshes with mixed quads and triangles, leading to a permutation failure, which assumes a single cell type. This does not happen in the x86-64 based images.


### How to reproduce the bug

Run `python3 -m pytest python/test/unit/mesh/test_higher_order_mesh.py` on an ARM-based machine any recently released `v0.7.x` containers with DOLFINx `v0.7.x` source code checked out. 

### Minimal Example (Python)

_No response_

### Output (Python)

```
========================================================= test session starts ==========================================================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /shared/dolfinx
plugins: xdist-3.5.0
collected 0 items

======================================================== no tests ran in 0.01s =========================================================
ERROR: file or directory not found: python/unit/test/mesh/test_higher_order_mesh.py

root@d3533dad2e3b:/shared/dolfinx# python3 -m pytest python/test/unit/mesh/test_higher_order_mesh.py
========================================================= test session starts ==========================================================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /shared/dolfinx/python
configfile: setup.cfg
plugins: xdist-3.5.0
collected 53 items

python/test/unit/mesh/test_higher_order_mesh.py ...........................x......xx.......F.F.....x.                            [100%]

=============================================================== FAILURES ===============================================================
___________________________________________________ test_gmsh_input_2d[cell_type1-1] ___________________________________________________

order = 1, cell_type = <CellType.quadrilateral: -4>

    @pytest.mark.skip_in_parallel
    @pytest.mark.parametrize('order', range(1, 4))
    @pytest.mark.parametrize('cell_type', [CellType.triangle, CellType.quadrilateral])
    def test_gmsh_input_2d(order, cell_type):
        try:
            import gmsh
        except ImportError:
            pytest.skip()
        res = 0.2
        gmsh.initialize()
        gmsh.option.setNumber("Mesh.CharacteristicLengthMin", res)
        gmsh.option.setNumber("Mesh.CharacteristicLengthMax", res)

        if cell_type == CellType.quadrilateral:
            gmsh.option.setNumber("Mesh.Algorithm", 2 if order == 2 else 5)

        gmsh.model.occ.addSphere(0, 0, 0, 1, tag=1)
        gmsh.model.occ.synchronize()

        gmsh.model.mesh.generate(2)
        if cell_type == CellType.quadrilateral:
            gmsh.model.mesh.recombine()
        gmsh.model.mesh.setOrder(order)
        idx, points, _ = gmsh.model.mesh.getNodes()
        points = points.reshape(-1, 3)
        idx -= 1
        srt = np.argsort(idx)
        assert np.all(idx[srt] == np.arange(len(idx)))
        x = points[srt]

        element_types, element_tags, node_tags = gmsh.model.mesh.getElements(dim=2)
        name, dim, order, num_nodes, local_coords, num_first_order_nodes = gmsh.model.mesh.getElementProperties(
            element_types[0])

        cells = node_tags[0].reshape(-1, num_nodes) - 1
        if cell_type == CellType.triangle:
            gmsh_cell_id = gmsh.model.mesh.getElementType("triangle", order)
        elif cell_type == CellType.quadrilateral:
            gmsh_cell_id = gmsh.model.mesh.getElementType("quadrangle", order)
        gmsh.finalize()

>       cells = cells[:, cell_perm_array(cell_type, cells.shape[1])]

python/test/unit/mesh/test_higher_order_mesh.py:638:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cell_type = <CellType.quadrilateral: -4>, num_nodes = 3

    def cell_perm_array(cell_type: CellType, num_nodes: int) -> typing.List[int]:
        """The permutation array for permuting Gmsh ordering to DOLFINx ordering.

        Args:
            cell_type: DOLFINx cell type.
            num_nodes: Number of nodes in the cell.

        Returns:
            An array ``p`` such that ``a_dolfinx[i] = a_gmsh[p[i]]``.

        """
>       return _cpp.io.perm_gmsh(cell_type, num_nodes)
E       RuntimeError: Higher order Gmsh quadrilateral not supported

/usr/local/dolfinx-real/lib/python3.10/dist-packages/dolfinx/io/gmshio.py:78: RuntimeError
--------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Info    : Meshing 1D...
Info    : [ 40%] Meshing curve 2 (Circle)
Info    : Done meshing 1D (Wall 6.2209e-05s, CPU 5.7e-05s)
Info    : Meshing 2D...
Info    : Meshing surface 1 (Sphere, Delaunay)
Info    : Done meshing 2D (Wall 0.00901912s, CPU 0.009019s)
Info    : 444 nodes 902 elements
Info    : Recombining 2D mesh...
Info    : Blossom: 1310 internal 0 closed
Info    : Blossom recombination completed (Wall 0.00780987s, CPU 0.007809s): 428 quads, 2 triangles, 0 invalid quads, 0 quads with Q < 0.1, avg Q = 0.744147, min Q = 0.113929
Info    : Done recombining 2D mesh (Wall 0.00785229s, CPU 0.007852s)
___________________________________________________ test_gmsh_input_2d[cell_type1-3] ___________________________________________________

order = 3, cell_type = <CellType.quadrilateral: -4>

    @pytest.mark.skip_in_parallel
    @pytest.mark.parametrize('order', range(1, 4))
    @pytest.mark.parametrize('cell_type', [CellType.triangle, CellType.quadrilateral])
    def test_gmsh_input_2d(order, cell_type):
        try:
            import gmsh
        except ImportError:
            pytest.skip()
        res = 0.2
        gmsh.initialize()
        gmsh.option.setNumber("Mesh.CharacteristicLengthMin", res)
        gmsh.option.setNumber("Mesh.CharacteristicLengthMax", res)

        if cell_type == CellType.quadrilateral:
            gmsh.option.setNumber("Mesh.Algorithm", 2 if order == 2 else 5)

        gmsh.model.occ.addSphere(0, 0, 0, 1, tag=1)
        gmsh.model.occ.synchronize()

        gmsh.model.mesh.generate(2)
        if cell_type == CellType.quadrilateral:
            gmsh.model.mesh.recombine()
        gmsh.model.mesh.setOrder(order)
        idx, points, _ = gmsh.model.mesh.getNodes()
        points = points.reshape(-1, 3)
        idx -= 1
        srt = np.argsort(idx)
        assert np.all(idx[srt] == np.arange(len(idx)))
        x = points[srt]

        element_types, element_tags, node_tags = gmsh.model.mesh.getElements(dim=2)
        name, dim, order, num_nodes, local_coords, num_first_order_nodes = gmsh.model.mesh.getElementProperties(
            element_types[0])

        cells = node_tags[0].reshape(-1, num_nodes) - 1
        if cell_type == CellType.triangle:
            gmsh_cell_id = gmsh.model.mesh.getElementType("triangle", order)
        elif cell_type == CellType.quadrilateral:
            gmsh_cell_id = gmsh.model.mesh.getElementType("quadrangle", order)
        gmsh.finalize()

>       cells = cells[:, cell_perm_array(cell_type, cells.shape[1])]

python/test/unit/mesh/test_higher_order_mesh.py:638:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cell_type = <CellType.quadrilateral: -4>, num_nodes = 10

    def cell_perm_array(cell_type: CellType, num_nodes: int) -> typing.List[int]:
        """The permutation array for permuting Gmsh ordering to DOLFINx ordering.

        Args:
            cell_type: DOLFINx cell type.
            num_nodes: Number of nodes in the cell.

        Returns:
            An array ``p`` such that ``a_dolfinx[i] = a_gmsh[p[i]]``.

        """
>       return _cpp.io.perm_gmsh(cell_type, num_nodes)
E       RuntimeError: Higher order Gmsh quadrilateral not supported

/usr/local/dolfinx-real/lib/python3.10/dist-packages/dolfinx/io/gmshio.py:78: RuntimeError
--------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Info    : Meshing 1D...
Info    : [ 40%] Meshing curve 2 (Circle)
Info    : Done meshing 1D (Wall 8.0583e-05s, CPU 8.2e-05s)
Info    : Meshing 2D...
Info    : Meshing surface 1 (Sphere, Delaunay)
Info    : Done meshing 2D (Wall 0.00908258s, CPU 0.009082s)
Info    : 444 nodes 902 elements
Info    : Recombining 2D mesh...
Info    : Blossom: 1310 internal 0 closed
Info    : Blossom recombination completed (Wall 0.00764929s, CPU 0.007649s): 428 quads, 2 triangles, 0 invalid quads, 0 quads with Q < 0.1, avg Q = 0.744147, min Q = 0.113929
Info    : Done recombining 2D mesh (Wall 0.00768946s, CPU 0.007689s)
Info    : Meshing order 3 (curvilinear on)...
Info    : [  0%] Meshing curve 1 order 3
Info    : [ 20%] Meshing curve 2 order 3
Info    : [ 40%] Meshing curve 3 order 3
Info    : [ 60%] Meshing surface 1 order 3
Info    : [ 80%] Meshing volume 1 order 3
Info    : Surface mesh: worst distortion = 0.988495 (0 elements in ]0, 0.2]); worst gamma = 0.782841
Info    : Done meshing order 3 (Wall 0.00362704s, CPU 0.003627s)
=========================================================== warnings summary ===========================================================
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
  /usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56: DeprecationWarning: attach_operators_from_hash_data deprecated, please use UFLObject instead.
    warnings.warn("attach_operators_from_hash_data deprecated, please use UFLObject instead.", DeprecationWarning)

test/unit/mesh/test_higher_order_mesh.py: 32 warnings
  /usr/local/lib/python3.10/dist-packages/ufl/core/expr.py:275: DeprecationWarning: Expr.ufl_domain() is deprecated, please use extract_unique_domain(expr) instead.
    warnings.warn("Expr.ufl_domain() is deprecated, please "

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= short test summary info ========================================================
FAILED python/test/unit/mesh/test_higher_order_mesh.py::test_gmsh_input_2d[cell_type1-1] - RuntimeError: Higher order Gmsh quadrilateral not supported
FAILED python/test/unit/mesh/test_higher_order_mesh.py::test_gmsh_input_2d[cell_type1-3] - RuntimeError: Higher order Gmsh quadrilateral not supported
========================================= 2 failed, 47 passed, 4 xfailed, 40 warnings in 5.43s =========================================
root@d3533dad2e3b:/shared/dolfinx# python3 -m pytest python/test/unit/mesh/test_higher_order_mesh.py
========================================================= test session starts ==========================================================
platform linux -- Python 3.10.12, pytest-7.4.3, pluggy-1.3.0
rootdir: /shared/dolfinx/python
configfile: setup.cfg
plugins: xdist-3.5.0
collected 53 items

python/test/unit/mesh/test_higher_order_mesh.py ...........................x......xx.......F.F.....x.                            [100%]

=============================================================== FAILURES ===============================================================
___________________________________________________ test_gmsh_input_2d[cell_type1-1] ___________________________________________________

order = 1, cell_type = <CellType.quadrilateral: -4>

    @pytest.mark.skip_in_parallel
    @pytest.mark.parametrize('order', range(1, 4))
    @pytest.mark.parametrize('cell_type', [CellType.triangle, CellType.quadrilateral])
    def test_gmsh_input_2d(order, cell_type):
        try:
            import gmsh
        except ImportError:
            pytest.skip()
        res = 0.2
        gmsh.initialize()
        gmsh.option.setNumber("Mesh.CharacteristicLengthMin", res)
        gmsh.option.setNumber("Mesh.CharacteristicLengthMax", res)

        if cell_type == CellType.quadrilateral:
            gmsh.option.setNumber("Mesh.Algorithm", 2 if order == 2 else 5)

        gmsh.model.occ.addSphere(0, 0, 0, 1, tag=1)
        gmsh.model.occ.synchronize()

        gmsh.model.mesh.generate(2)
        if cell_type == CellType.quadrilateral:
            gmsh.model.mesh.recombine()
        gmsh.model.mesh.setOrder(order)
        idx, points, _ = gmsh.model.mesh.getNodes()
        points = points.reshape(-1, 3)
        idx -= 1
        srt = np.argsort(idx)
        assert np.all(idx[srt] == np.arange(len(idx)))
        x = points[srt]

        element_types, element_tags, node_tags = gmsh.model.mesh.getElements(dim=2)
        name, dim, order, num_nodes, local_coords, num_first_order_nodes = gmsh.model.mesh.getElementProperties(
            element_types[0])

        cells = node_tags[0].reshape(-1, num_nodes) - 1
        if cell_type == CellType.triangle:
            gmsh_cell_id = gmsh.model.mesh.getElementType("triangle", order)
        elif cell_type == CellType.quadrilateral:
            gmsh_cell_id = gmsh.model.mesh.getElementType("quadrangle", order)
        gmsh.finalize()

>       cells = cells[:, cell_perm_array(cell_type, cells.shape[1])]

python/test/unit/mesh/test_higher_order_mesh.py:638:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cell_type = <CellType.quadrilateral: -4>, num_nodes = 3

    def cell_perm_array(cell_type: CellType, num_nodes: int) -> typing.List[int]:
        """The permutation array for permuting Gmsh ordering to DOLFINx ordering.

        Args:
            cell_type: DOLFINx cell type.
            num_nodes: Number of nodes in the cell.

        Returns:
            An array ``p`` such that ``a_dolfinx[i] = a_gmsh[p[i]]``.

        """
>       return _cpp.io.perm_gmsh(cell_type, num_nodes)
E       RuntimeError: Higher order Gmsh quadrilateral not supported

/usr/local/dolfinx-real/lib/python3.10/dist-packages/dolfinx/io/gmshio.py:78: RuntimeError
--------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Info    : Meshing 1D...
Info    : [ 40%] Meshing curve 2 (Circle)
Info    : Done meshing 1D (Wall 5.8333e-05s, CPU 5.9e-05s)
Info    : Meshing 2D...
Info    : Meshing surface 1 (Sphere, Delaunay)
Info    : Done meshing 2D (Wall 0.0091295s, CPU 0.00913s)
Info    : 444 nodes 902 elements
Info    : Recombining 2D mesh...
Info    : Blossom: 1310 internal 0 closed
Info    : Blossom recombination completed (Wall 0.00764142s, CPU 0.007641s): 428 quads, 2 triangles, 0 invalid quads, 0 quads with Q < 0.1, avg Q = 0.744147, min Q = 0.113929
Info    : Done recombining 2D mesh (Wall 0.00768108s, CPU 0.007681s)
___________________________________________________ test_gmsh_input_2d[cell_type1-3] ___________________________________________________

order = 3, cell_type = <CellType.quadrilateral: -4>

    @pytest.mark.skip_in_parallel
    @pytest.mark.parametrize('order', range(1, 4))
    @pytest.mark.parametrize('cell_type', [CellType.triangle, CellType.quadrilateral])
    def test_gmsh_input_2d(order, cell_type):
        try:
            import gmsh
        except ImportError:
            pytest.skip()
        res = 0.2
        gmsh.initialize()
        gmsh.option.setNumber("Mesh.CharacteristicLengthMin", res)
        gmsh.option.setNumber("Mesh.CharacteristicLengthMax", res)

        if cell_type == CellType.quadrilateral:
            gmsh.option.setNumber("Mesh.Algorithm", 2 if order == 2 else 5)

        gmsh.model.occ.addSphere(0, 0, 0, 1, tag=1)
        gmsh.model.occ.synchronize()

        gmsh.model.mesh.generate(2)
        if cell_type == CellType.quadrilateral:
            gmsh.model.mesh.recombine()
        gmsh.model.mesh.setOrder(order)
        idx, points, _ = gmsh.model.mesh.getNodes()
        points = points.reshape(-1, 3)
        idx -= 1
        srt = np.argsort(idx)
        assert np.all(idx[srt] == np.arange(len(idx)))
        x = points[srt]

        element_types, element_tags, node_tags = gmsh.model.mesh.getElements(dim=2)
        name, dim, order, num_nodes, local_coords, num_first_order_nodes = gmsh.model.mesh.getElementProperties(
            element_types[0])

        cells = node_tags[0].reshape(-1, num_nodes) - 1
        if cell_type == CellType.triangle:
            gmsh_cell_id = gmsh.model.mesh.getElementType("triangle", order)
        elif cell_type == CellType.quadrilateral:
            gmsh_cell_id = gmsh.model.mesh.getElementType("quadrangle", order)
        gmsh.finalize()

>       cells = cells[:, cell_perm_array(cell_type, cells.shape[1])]

python/test/unit/mesh/test_higher_order_mesh.py:638:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cell_type = <CellType.quadrilateral: -4>, num_nodes = 10

    def cell_perm_array(cell_type: CellType, num_nodes: int) -> typing.List[int]:
        """The permutation array for permuting Gmsh ordering to DOLFINx ordering.

        Args:
            cell_type: DOLFINx cell type.
            num_nodes: Number of nodes in the cell.

        Returns:
            An array ``p`` such that ``a_dolfinx[i] = a_gmsh[p[i]]``.

        """
>       return _cpp.io.perm_gmsh(cell_type, num_nodes)
E       RuntimeError: Higher order Gmsh quadrilateral not supported

/usr/local/dolfinx-real/lib/python3.10/dist-packages/dolfinx/io/gmshio.py:78: RuntimeError
--------------------------------------------------------- Captured stdout call ---------------------------------------------------------
Info    : Meshing 1D...
Info    : [ 40%] Meshing curve 2 (Circle)
Info    : Done meshing 1D (Wall 5.9166e-05s, CPU 5.5e-05s)
Info    : Meshing 2D...
Info    : Meshing surface 1 (Sphere, Delaunay)
Info    : Done meshing 2D (Wall 0.00910337s, CPU 0.009093s)
Info    : 444 nodes 902 elements
Info    : Recombining 2D mesh...
Info    : Blossom: 1310 internal 0 closed
Info    : Blossom recombination completed (Wall 0.00771225s, CPU 0.007712s): 428 quads, 2 triangles, 0 invalid quads, 0 quads with Q < 0.1, avg Q = 0.744147, min Q = 0.113929
Info    : Done recombining 2D mesh (Wall 0.00775238s, CPU 0.007752s)
Info    : Meshing order 3 (curvilinear on)...
Info    : [  0%] Meshing curve 1 order 3
Info    : [ 20%] Meshing curve 2 order 3
Info    : [ 40%] Meshing curve 3 order 3
Info    : [ 60%] Meshing surface 1 order 3
Info    : [ 80%] Meshing volume 1 order 3
Info    : Surface mesh: worst distortion = 0.988495 (0 elements in ]0, 0.2]); worst gamma = 0.782841
Info    : Done meshing order 3 (Wall 0.00355042s, CPU 0.003521s)
=========================================================== warnings summary ===========================================================
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
../../usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56
  /usr/local/lib/python3.10/dist-packages/ufl/core/ufl_type.py:56: DeprecationWarning: attach_operators_from_hash_data deprecated, please use UFLObject instead.
    warnings.warn("attach_operators_from_hash_data deprecated, please use UFLObject instead.", DeprecationWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= short test summary info ========================================================
FAILED python/test/unit/mesh/test_higher_order_mesh.py::test_gmsh_input_2d[cell_type1-1] - RuntimeError: Higher order Gmsh quadrilateral not supported
FAILED python/test/unit/mesh/test_higher_order_mesh.py::test_gmsh_input_2d[cell_type1-3] - RuntimeError: Higher order Gmsh quadrilateral not supported
```

### Version

0.7.2

### DOLFINx git commit

_No response_

### Installation

_No response_

### Additional information

_No response_

--- a/python/test/unit/mesh/test_higher_order_mesh.py
+++ b/python/test/unit/mesh/test_higher_order_mesh.py
@@ -607,7 +607,9 @@ def test_gmsh_input_2d(order, cell_type):
     gmsh.option.setNumber("Mesh.CharacteristicLengthMax", res)
 
     if cell_type == CellType.quadrilateral:
-        gmsh.option.setNumber("Mesh.Algorithm", 2 if order == 2 else 5)
+        gmsh.option.setNumber("Mesh.Algorithm", 2)
+        # Force mesh to have no triangles
+        gmsh.option.setNumber("Mesh.RecombinationAlgorithm", 3)
 
     gmsh.model.occ.addSphere(0, 0, 0, 1, tag=1)
     gmsh.model.occ.synchronize()

